### PR TITLE
Getting Started with k8s: Day 2

### PR DESCRIPTION
**Pull request by spara**

### CODE REVIEW

File: content/blog/getting-started-with-k8s-part6/index.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+meta_desc: "How to use infrastructure as code for day 2 maintenance tasks in Kubernetes."
+meta_image: day_2.png
+authors:
+    - sophia-parafina
+tags:
+    - kubernetes
+    - day 2
+---
+
+Your application has left the dev stage, passed the testing stage, and is now in production. As a developer, you might think that you’ve passed your responsibilities to operators. However, DevOps is a collaborative effort between developers and operators to build and maintain applications using shared techniques and processes. Day 2 activities are tasks for maintaining and improving applications in production.
```
Comment (lblackstone) [inline content/blog/getting-started-with-k8s-part6/index.md:13]:
```suggestion
Your application made it out of the dev stage, passed the testing stage, and arrived in production. As a developer, you might think that it's an ops problem now. However, DevOps is a collaborative effort between developers and operators to build and maintain applications using shared techniques and processes, often called "Day 2" activities.
```

File: content/blog/getting-started-with-k8s-part6/index.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+<!--more-->
+
+Day 2 tasks fall into three categories. The first category is managing cluster components. It involves managing nodes, users and groups, ingress and egress, networking, storage, and [operators](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/). The second category is changing cluster components, including setting optimal resource quota, reclaim resources, scale and tune clusters, and updating clusters and use [CRDs](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions). The third category of tasks is cluster monitoring. Cluster monitoring requires [cluster logging](https://kubernetes.io/docs/tasks/debug-application-cluster/), cluster monitoring involves collecting and analyzing cluster metrics from the logging data, and monitoring the health of the cluster.
```
Comment (lblackstone) [inline content/blog/getting-started-with-k8s-part6/index.md:17]:
```suggestion
Day 2 tasks fall into three categories. The first category is managing cluster components. It involves managing nodes, users and groups, ingress and egress, networking, storage, and [operators](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/). The second category is changing cluster components, including setting optimal resource quota, reclaiming unneeded resources, scaling, tuning and updating clusters, and managing [CRDs](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions). The third category of tasks is cluster monitoring and [logging](https://kubernetes.io/docs/tasks/debug-application-cluster/), which involves collecting and analyzing metrics, and monitoring the health of the cluster.
```

File: content/blog/getting-started-with-k8s-part6/index.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+Anyone of these tasks in these three categories would require more coverage than a single article. Let’s take a look at how we can use infrastructure as code to address some of these tasks.
```
Comment (lblackstone) [inline content/blog/getting-started-with-k8s-part6/index.md:19]:
```suggestion
Although this subject is much deeper than we can cover in a single article, let’s take a look at how infrastructure as code can simplify some of these tasks.
```

File: content/blog/getting-started-with-k8s-part6/index.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+## Managing secrets
+
+In addition to managing users and groups, managing secrets such as credentials is an important task. The Pulumi CLI lets you rotate secrets providers for a stack. This particularly important when people leave an organization or if you want to change to a different secrets provider, e.g., from Vault to AWS KMS. Also, we can copy secrets from one stack to another. A frequent use of this capability is to copy database account credentials from one stack to another. You can read more about [managing secrets]({{< relref "/blog/managing-secrets-with-pulumi" >}}) and the using the [CLI]({{< relref "blog/rotating-secret-providers" >}}) to rotate providers on the Pulumi blog.
```
Comment (lblackstone) [inline content/blog/getting-started-with-k8s-part6/index.md:23]:
```suggestion
In addition to managing users and groups, managing secrets such as credentials is an important task. The Pulumi CLI lets you rotate the secrets provider for a stack. This is particularly important when people leave an organization or if you want to change to a different secrets provider, e.g., from Vault to AWS KMS. Also, we can copy secrets from one stack to another. A frequent use of this capability is to copy database account credentials from one stack to another. You can read more about [managing secrets]({{< relref "/blog/managing-secrets-with-pulumi" >}}) and the using the [CLI]({{< relref "blog/rotating-secret-providers" >}}) to rotate providers on the Pulumi blog.
```

File: content/blog/getting-started-with-k8s-part6/index.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+In addition to zero downtime cluster migrations, we can also control scaling. The following tutorial shows how to stage an application rollout from dev to testing controlled by performance collecting metrics. In this [tutorial]({{< relref "/docs/tutorials/kubernetes/p8s-rollout" >}}), we move a three replica canary to a ten replica staging deployment. Replicas are added based on a P90 response time, i.e., 90% of requests are processed within 2.75 seconds. The example uses [Prometheus](https://prometheus.io/) to collect the response times and a [utility class](https://github.com/pulumi/examples/blob/master/kubernetes-ts-staged-rollout-with-prometheus/util.ts) to perform the health check. You can clone the [code on Github](https://github.com/pulumi/examples/tree/master/kubernetes-ts-staged-rollout-with-prometheus).
```
Comment (lblackstone) [inline content/blog/getting-started-with-k8s-part6/index.md:29]:
```suggestion
In addition to zero downtime cluster migrations, we can also control scaling. The following tutorial shows how to stage an application rollout from dev to testing controlled by performance collecting metrics. In this [tutorial]({{< relref "/docs/tutorials/kubernetes/p8s-rollout" >}}), we move a three replica canary to a ten replica staging deployment. Replicas are added based on a P90 response time, i.e., 90% of requests are processed within 2.75 seconds. The example uses [Prometheus](https://prometheus.io/) to collect the response times and a [utility class](https://github.com/pulumi/examples/blob/master/kubernetes-ts-staged-rollout-with-prometheus/util.ts) to perform the health check. Check out the complete [code on Github](https://github.com/pulumi/examples/tree/master/kubernetes-ts-staged-rollout-with-prometheus).
```

File: content/blog/getting-started-with-k8s-part6/index.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+Pulumi also lets you update your Kubernetes resources in real-time with Pulumi watch. You can add more replicas to deployment by saving changes to the stack. Adding a Service to make an application available outside the cluster with a load balancer is instantaneous. Watch how you can update your application.
```
Comment (lblackstone) [inline content/blog/getting-started-with-k8s-part6/index.md:31]:
```suggestion
Pulumi also lets you keep an eye on your Kubernetes resources in real-time with Pulumi watch. Results are updated live as you make changes like adding replicas or exposing an application with a load balancer.
```

File: content/blog/getting-started-with-k8s-part6/index.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+{{< youtube "X96EMLi8uJY?rel=0" >}}
+
+## Monitoring
+
+There are many [logging and monitoring](https://kubernetes.io/docs/tasks/debug-application-cluster/) solutions available for Kubernetes. In the previous section, we used Prometheus for generating metrics. Pulumi supports logging and monitoring tools to assist with cluster and application management.
+
+Kubernetes Cluster Services provide logging and monitoring at the cluster level or a subset of apps and workloads. Tutorials for setting up logging and monitoring for Kubernetes on [AWS, Azure, and GCP]({{< relref "/docs/guides/crosswalk/kubernetes/cluster-services" >}}). The tutorial also shows how to configure [DataDog](https://www.datadoghq.com/) to log and monitor applications. The [example code](https://github.com/pulumi/kubernetes-guides/tree/master/general-cluster-services/datadog-daemonset) is on Github.
```
Comment (lblackstone) [inline content/blog/getting-started-with-k8s-part6/index.md:39]:
```suggestion
Kubernetes Cluster Services provide logging and monitoring at the cluster level or a subset of apps and workloads. Check out our tutorials for setting up logging and monitoring for Kubernetes on [AWS, Azure, and GCP]({{< relref "/docs/guides/crosswalk/kubernetes/cluster-services" >}}). The tutorial also shows how to configure [DataDog](https://www.datadoghq.com/) to log and monitor applications. The [example code](https://github.com/pulumi/kubernetes-guides/tree/master/general-cluster-services/datadog-daemonset) is on Github.
```

File: content/blog/getting-started-with-k8s-part6/index.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+Pulumi also includes tools for managing resources. [Kubernetes Query](https://github.com/pulumi/pulumi-query-kubernetes) supports both batch and streaming queries that lets you ask questions such as, “How many Pods are in a Node?” Batch queries run to completion and return results based on the current configuration. Streaming queries continuously watch the cluster and take action based on criteria; for example, a streaming query can watch storage and warn you if the storage is 80% full. You can read about typical use cases on the Pulumi [blog]({{< relref "/blog/query-kubernetes" >}}).
+
+## Summary
+
+Maintaining a Kubernetes cluster and modern applications can be varied and complex. However, infrastructure as code lets developers and operators use a common language to manage the cluster and resources. Pulumi has a rich toolset to help you accomplish Day 2 tasks, whether managing cluster components such as secrets, changing cluster components to scale applications, or log and monitor your cluster and resources.
```
Comment (lblackstone) [inline content/blog/getting-started-with-k8s-part6/index.md:45]:
```suggestion
Maintaining a Kubernetes cluster and modern applications can be varied and complex. However, infrastructure as code lets developers and operators use a common language to manage the cluster and resources. Pulumi has a rich toolset to help you accomplish Day 2 tasks, whether managing cluster components such as secrets, changing cluster components to scale applications, or logging and monitoring your cluster and resources.
```

File: content/blog/getting-started-with-k8s-part6/index.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+Kubernetes Cluster Services provide logging and monitoring at the cluster level or a subset of apps and workloads. Tutorials for setting up logging and monitoring for Kubernetes on [AWS, Azure, and GCP]({{< relref "/docs/guides/crosswalk/kubernetes/cluster-services" >}}). The tutorial also shows how to configure [DataDog](https://www.datadoghq.com/) to log and monitor applications. The [example code](https://github.com/pulumi/kubernetes-guides/tree/master/general-cluster-services/datadog-daemonset) is on Github.
+
+Pulumi also includes tools for managing resources. [Kubernetes Query](https://github.com/pulumi/pulumi-query-kubernetes) supports both batch and streaming queries that lets you ask questions such as, “How many Pods are in a Node?” Batch queries run to completion and return results based on the current configuration. Streaming queries continuously watch the cluster and take action based on criteria; for example, a streaming query can watch storage and warn you if the storage is 80% full. You can read about typical use cases on the Pulumi [blog]({{< relref "/blog/query-kubernetes" >}}).
```
Comment (metral) [inline content/blog/getting-started-with-k8s-part6/index.md:41]: Do we want to mention kquery? It's not actively maintained, and I don't believe it really fits here. Do you believe it fits inline with the day2 story?

Comment (spara) [reply content/blog/getting-started-with-k8s-part6/index.md:41]: I agree it should be cut if we're not maintaining it.

**Pull request merged**